### ISSUE TITLE
Prevent simultaneous pulls of Singularity images by toil-cwl-runner: prefetch them all

### ISSUE THREAD

**Issue by stevekm** [DataBiosphere/toil#3817]
Toil / cwltool should be updated to pull all Docker / Singularity containers by default before a pipeline starts running. 

If you are running a pipeline that launches multiple parallel scatter jobs, and the container for the jobs cannot be found locally, the results can be disastrous. 

This is an example what happened when I tried to run a CWL pipeline with Toil on our HPC (using LSF);

```
kellys5  33461  2.8  0.0 978332 57072 ?        Sl   16:18   1:36 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33462  2.8  0.0 978332 59232 ?        Sl   16:18   1:35 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33464  2.9  0.0 912796 58996 ?        Sl   16:18   1:40 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33465  3.0  0.0 912796 55508 ?        Sl   16:18   1:43 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33466  3.1  0.0 912796 53428 ?        Sl   16:18   1:47 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33469  3.1  0.0 912796 56456 ?        Sl   16:18   1:45 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33470  3.1  0.0 912796 54580 ?        Sl   16:18   1:44 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  33471  3.0  0.0 912796 55364 ?        Sl   16:18   1:41 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  34773  2.9  0.0 912796 57896 ?        Sl   16:19   1:36 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  35503  2.8  0.0 978332 56740 ?        Sl   16:19   1:34 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  35504  2.9  0.0 978332 55564 ?        Sl   16:19   1:37 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
kellys5  35505  2.8  0.0 912796 54004 ?        Sl   16:19   1:33 singularity pull --force --name stevekm_some-container:1.2.3.sif docker://stevekm/some-container:1.2.3
```

Dozens of jobs were launched, each one of them attempting to do a `singularity pull` for the same image at the same time. This ended up destabilizing the HPC filesystem for all users. 

Pulling Singularity containers in parallel like this is also advised against by the Singularity docs because it is highly likely to crash, leave a corrupted container, or otherwise error out;

https://sylabs.io/guides/3.6/user-guide/build_env.html#cache-folders
> Warning
> 
> If you are not certain that your $HOME or SINGULARITY_CACHEDIR filesytems support atomic rename, do not run Singularity in parallel using remote container URLs. Instead use singularity pull to create a local SIF image, and then run this SIF image in a parallel step.

An example error message I commonly saw was this;

```
FATAL:   could not open image /toil_runs/1632937855/work/cda001d6cc1d55f499cb751325575255/71fc/c89a/some_container:1.2.3.sif: failed to retrieve path for /toil_runs/1632937855/work/cda001d6cc1d55f499cb751325575255/71fc/c89a/some_container:1.2.3.sif: lstat /toil_runs/1632937855/work/cda001d6cc1d55f499cb751325575255/71fc/c89a/some_container:1.2.3.sif: no such file or directory
```

As a naive Toil user, I had no clue that such issues could arise just by running a pipeline that worked fine in other contexts. I ultimately found some resolution by manually doing `singularity pull` on all the known containers with env vars `SINGULARITY_CACHEDIR`, `SINGULARITY_TMPDIR`, `SINGULARITY_PULLDIR`, and `CWL_SINGULARITY_CACHE=SINGULARITY_PULLDIR` properly set and exported from the host environment to child LSF jobs. 

However, there appears to already be a tool developed for this specific purpose:
https://github.com/common-workflow-language/cwl-utils/#pull-the-all-referenced-software-container-images

Since this tool is already in the CWL codebase, and since the lack of this behavior can cause severe problems in running a CWL pipeline with Toil, it seems that this should just be merged into Toil itself and should become the default behavior. This will likely save users a huge amount of time and headaches and might even prevent some servers from being crashed.

┆Issue is synchronized with this [Jira Story](https://ucsc-cgl.atlassian.net/browse/TOIL-1042)
┆Issue Number: TOIL-1042

**Comment by adamnovak** [comment]
The fiddly parts here are going to be:

1. What do we do if workers have their own Singularity image caches because they have their own filesystems? We wouldn't be able to just pull at the start on the leader. but we still might get multiple simultaneous pulls on an individual worker node.
2. How much of this can `cwltool` be responsible for? `toil-cwl-runner` never actually makes any Singularity calls, and doesn't provide a Singularity-calling interface that `cwltool` calls. Maybe `cwltool` has to be responsible for caching/locking/prefetching to protect against simultaneous pulls by multiple jobs or in multiple workflows?

`toil-vg` has its own code to solve this multiple-pulls-on-the-same-filesystem problem, but all it really does is make sure everyone is pulling into their own directory, and that somebody's pull eventually gets cached: https://github.com/vgteam/toil-vg/blob/9ef7175d110d8f82e63330fb275c05cd8c1c7f32/src/toil_vg/singularity.py#L176-L228

**Comment by mr-c** [comment]
>  could `cwltool` be induced to call it?

Sure, we could make https://github.com/common-workflow-language/cwltool/blob/aec33fcfa3459a90cbba8c88ebb991be94d21429/cwltool/singularity.py#L65 work with another API

>  Or does Toil need to check on every worker startup whether all the Singularity images the workflow will have are cached

Ideally, Toil would cache the images up front before scheduling jobs

**Comment by adamnovak** [comment]
I don't think we can cache images before sending the jobs to the nodes when there isn't a shared filesystem between the nodes, so we do fundamentally have to support multiple jobs on the same node trying to pull Singularity containers at once.

I think right now we have some special handling for `SINGULARITY_CACHEDIR` in Toil. We could add some machinery to set good `SINGULARITY_TMPDIR`/`SINGULARITY_PULLDIR` values.

But I think what we might really need is machinery in `cwltool` to deal with the fact that the Singularity caching system is not concurrent, and the user of `singularity` is not allowed to make multiple calls to `singularity pull` of the same image at the same time. `cwltool` ought to support a user invoking several workflows at the same time on the same machine without having all of them try to pull the container into the same place.

@mr-c is that interlock something you want to implement on the cwltool team side, or do we need to try and contribute it?
